File: mcp-server/fs-index-server/mount_discovery.go
package main

import (
	"log"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
)

// FilespaceMount represents a discovered LucidLink filespace mount.
type FilespaceMount struct {
	InstanceID string
	MountPoint string
	Name       string // parsed from lucid list output
	Port       int    // LucidLink client API port (from lucid list)
}

// discoverMounts uses `lucid list` to get instance IDs and ports, then
// `lucid --instance <id> status` to parse each mount point.
func discoverMounts(lucidBin string) []FilespaceMount {
	// Step 1: Get instance list (includes ports)
	out, err := exec.Command(lucidBin, "list").CombinedOutput()
	if err != nil {
		log.Printf("mount discovery: lucid list failed: %v (%s)", err, strings.TrimSpace(string(out)))
		return nil
	}

	instances := parseInstanceList(string(out))
	if len(instances) == 0 {
		log.Printf("mount discovery: no instances found in lucid list output")
		return nil
	}

	// Step 2: Get mount point for each instance
	var mounts []FilespaceMount
	for _, inst := range instances {
		mount := getInstanceMount(lucidBin, inst.id)
		if mount != nil {
			mount.Port = inst.port
			mounts = append(mounts, *mount)
		}
	}

	return mounts
}

// instanceInfo holds parsed data from a single lucid list row.
type instanceInfo struct {
	id   string
	name string
	port int
}

// parseInstanceList parses `lucid list` output to extract instance ID, name, and port.
// Example output:
//   INSTANCE ID        FILESPACE                    PORT        MODE
//   2045               connect-us.lucid-demo        9823        live
func parseInstanceList(output string) []instanceInfo {
	var result []instanceInfo
	lines := strings.Split(strings.TrimSpace(output), "\n")

	// Regex: instance_id  filespace_name  port  mode
	lineRe := regexp.MustCompile(`^\s*(\d+)\s+(\S+)\s+(\d+)\s+(\S+)`)

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "INSTANCE") || strings.HasPrefix(line, "---") {
			continue
		}
		if m := lineRe.FindStringSubmatch(line); len(m) > 3 {
			port, _ := strconv.Atoi(m[3])
			result = append(result, instanceInfo{
				id:   m[1],
				name: m[2],
				port: port,
			})
		}
	}
	return result
}

// getInstanceMount runs `lucid --instance <id> status` and parses the mount point.
func getInstanceMount(lucidBin, instanceID string) *FilespaceMount {
	out, err := exec.Command(lucidBin, "--instance", instanceID, "status").CombinedOutput()
	if err != nil {
		log.Printf("mount discovery: lucid --instance %s status failed: %v", instanceID, err)
		return nil
	}

	output := string(out)
	mountPoint := parseMountPoint(output)
	if mountPoint == "" {
		log.Printf("mount discovery: no mount point found for instance %s", instanceID)
		return nil
	}

	name := parseFilespaceName(output)
	if name == "" {
		name = instanceID
	}

	return &FilespaceMount{
		InstanceID: instanceID,
		MountPoint: mountPoint,
		Name:       name,
	}
}

// parseMountPoint extracts "Mount point: /path/to/mount" from lucid status output.
func parseMountPoint(output string) string {
	for _, line := range strings.Split(output, "\n") {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "Mount point:") {
			return strings.TrimSpace(strings.TrimPrefix(line, "Mount point:"))
		}
	}
	return ""
}

// parseFilespace name extracts "Filespace: name.domain" from lucid status output.
func parseFilespaceNameFromStatus(output string) string {
	for _, line := range strings.Split(output, "\n") {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "Filespace:") {
			return strings.TrimSpace(strings.TrimPrefix(line, "Filespace:"))
		}
	}
	return ""
}

// parseFilespace extracts a name from status output, trying multiple patterns.
func parseFilespaceName(output string) string {
	if name := parseFilespaceNameFromStatus(output); name != "" {
		return name
	}
	// Fallback: look for filespace.domain pattern anywhere
	re := regexp.MustCompile(`([a-zA-Z0-9_-]+\.[a-zA-Z0-9_-]+)`)
	if m := re.FindStringSubmatch(output); len(m) > 1 {
		return m[1]
	}
	return ""
}

File: mcp-server/src/audit-trail/mount-discovery.ts
/**
 * Discover mounted LucidLink filespaces via the lucid CLI.
 *
 * Uses `lucid list` to get instance IDs, names, and ports, then
 * `lucid --instance <id> status` to parse mount points.
 *
 * Mirrors the Go implementation in fs-index-server/mount_discovery.go.
 */
import { execFile } from "node:child_process";

export interface FilespaceMount {
  instanceId: string;
  name: string;
  mountPoint: string;
  port: number;
}

function exec(cmd: string, args: string[]): Promise<{ stdout: string; stderr: string }> {
  return new Promise((resolve, reject) => {
    execFile(cmd, args, { timeout: 10_000 }, (error, stdout, stderr) => {
      if (error) {
        reject(new Error(`${cmd} ${args.join(" ")} failed: ${error.message}\n${stderr}`));
      } else {
        resolve({ stdout, stderr });
      }
    });
  });
}

/**
 * Parse `lucid list` output.
 *
 * Example:
 *   INSTANCE ID        FILESPACE                    PORT        MODE
 *   2004               nab.lucid-demo               9823        live
 */
function parseInstanceList(output: string): Array<{ id: string; name: string; port: number }> {
  const result: Array<{ id: string; name: string; port: number }> = [];
  const lineRe = /^\s*(\d+)\s+(\S+)\s+(\d+)\s+(\S+)/;

  for (const line of output.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("INSTANCE") || trimmed.startsWith("---")) continue;
    const m = lineRe.exec(trimmed);
    if (m) {
      result.push({ id: m[1], name: m[2], port: parseInt(m[3], 10) });
    }
  }
  return result;
}

/**
 * Parse `lucid --instance <id> status` output for mount point and filespace name.
 */
function parseMountPoint(output: string): string | null {
  for (const line of output.split("\n")) {
    const trimmed = line.trim();
    if (trimmed.startsWith("Mount point:")) {
      return trimmed.replace("Mount point:", "").trim();
    }
  }
  return null;
}

function parseFilespaceName(output: string): string | null {
  for (const line of output.split("\n")) {
    const trimmed = line.trim();
    if (trimmed.startsWith("Filespace:")) {
      return trimmed.replace("Filespace:", "").trim();
    }
  }
  return null;
}

/**
 * Discover all mounted LucidLink filespaces.
 * Returns an array of mounts with instance ID, name, mount point, and port.
 */
export async function discoverMounts(lucidBin = "lucid"): Promise<FilespaceMount[]> {
  // Step 1: lucid list
  const { stdout: listOutput } = await exec(lucidBin, ["list"]);
  const instances = parseInstanceList(listOutput);

  if (instances.length === 0) {
    return [];
  }

  // Step 2: get mount point for each instance
  const mounts: FilespaceMount[] = [];
  for (const inst of instances) {
    try {
      const { stdout: statusOutput } = await exec(lucidBin, ["--instance", inst.id, "status"]);
      const mountPoint = parseMountPoint(statusOutput);
      if (mountPoint) {
        const name = parseFilespaceName(statusOutput) ?? inst.name;
        mounts.push({
          instanceId: inst.id,
          name,
          mountPoint,
          port: inst.port,
        });
      }
    } catch {
      // Skip instances that fail status check
    }
  }

  return mounts;
}
